accepted/future-releases/unified-collections/feature-specification.md

```diff
@@ -0,0 +1,802 @@
+    ```
+
+*   The collection is a map and the value type of any of the `leafElements` may
+    not be assigned to the map's value type.
+
+    ```dart
+    <int, int>{if (true) 1: "not int"} // Error.
+    ```
+
+*   A spread element in a list or set literal has a static type that is not
+    `dynamic` or a subtype of `Iterable<Object>`.
+
+*   A spread element in a list or set has a static type that implements
+    `Iterable<T>` for some `T` and `T` is not assignable to the element type of
+    the list.
+
+*   A spread element in a map literal has a static type that is not `dynamic` or
+    a subtype of `Map<Object, Object>`.
+
+*   If a map spread element's static type implements `Map<K, V>` for some `K`
+    and `V` and `K` is not assignable to the key type of the map or `V` is not
+    assignable to the value type of the map.
+
+*   The type of the condition expression in an `if` element may not be assigned
+    to `bool`.
+
+    ```dart
+    [if ("not bool") 1] // Error.
+    ```
+
+*   The type of the iterator expression in a synchronous `for-in` element may
+    not be assigned to `Iterable<T>` for some type `T`. Otherwise, the *iterable
+    type* of the iterator is `T`.
+
+    ```dart
+    [for (var i in "not iterable") i] // Error.
+    ```
+
+*   The iterable type of the iterator in a synchronous `for-in` element may not
+    be assigned to the `for-in` variable's type.
+
+    ```dart
+    [for (int i in ["not", "int"]) i] // Error.
+    ```
+
+*   The type of the stream expression in an asynchronous `await for-in`
+    element may not be assigned to `Stream<T>` for some type `T`. Otherwise,
+    the *stream type* of the stream is `T`.
+
+    ```dart
+    [await for (var i in "not stream") i] // Error.
+    ```
+
+*   The stream type of the iterator in an asynchronous `await for-in` element
+    may not be assigned to the `for-in` variable's type.
+
+    ```dart
+    [await for (int i in Stream.fromIterable(["not", "int"])) i] // Error.
+    ```
+
+*   `await for` is used when the function immediately enclosing the collection
+    literal is not asynchronous.
+
+    ```dart
+    main() {
+      [await for (_ in stream)]; // Error.
+    }
+    ```
+
+*   `await` is used before a C-style `for` element. `await` can only be used
+    with `for-in` loops.
+
+    ```dart
+    main() {
+      [await for (;;)]; // Error.
+    }
+    ```
+
+*   The type of the condition expression (the second clause) in a C-style `for`
+    element may not be assigned to `bool`.
+
+    ```dart
+    [for (; "not bool";) 1] // Error.
+    ```
+
+## Constant Semantics
+
+The runtime semantics below can also be used to determine the compile-time value
+of a collection literal marked `const`, with a few modifications:
+
+*   A `listLiteral` is constant if it occurs in a constant context or if it is
+    directly prefixed by `const`. If so, then its `elements` must all be
+    constant elements, or at least potentially constant elements.
```

Comment:
This just means that they must be potentially constant (because all constant expressions are also potentially constant).

But it would be a rather big step to take if we allow potentially constant expressions (including the ones that are not constant) in constant list literals: This would allow us to create an unbounded number of constant lists based on a single constant list literal.

Do we want to allow this expansion of the expressive power of constant expressions?

Reply:
That definitely wasn't my intent. I lifted most of this prose from @lrhn . I'll dial it back for now unless Lasse explains why he wrote it this way. (Or maybe I just messed it up when I moved it over and reworded some things.)